xplique/commons/operators.py

```diff
@@ -0,0 +1,140 @@
+
+
+@tf.function
+def predictions_operator(model: Callable,
+                         inputs: tf.Tensor,
+                         targets: tf.Tensor) -> tf.Tensor:
+    """
+    Compute predictions scores, only for the label class, for a batch of samples.
+
+    Parameters
+    ----------
+    model
+        Model used for computing predictions.
+    inputs
+        Input samples to be explained.
+    targets
+        One-hot encoded labels or regression target (e.g {+1, -1}), one for each sample.
+
+    Returns
+    -------
+    scores
+        Predictions scores computed, only for the label class.
+    """
+    scores = tf.reduce_sum(model(inputs) * targets, axis=-1)
+    return scores
+
+@tf.function
+def binary_segmentation_operator(model: Callable,
+                                 inputs: tf.Tensor,
+                                 targets: tf.Tensor) -> tf.Tensor:
+    """
+    Compute the segmentation score for a batch of samples.
+
+    Parameters
+    ----------
+    model
+        Model used for computing predictions.
+    inputs
+        Input samples to be explained.
+    targets
+        One-hot encoded labels or regression target (e.g {+1, -1}), one for each sample.
+
+    Returns
+    -------
+    scores
+        Segmentation scores computed.
+    """
+    scores = tf.reduce_sum(model(inputs) * targets, axis=(1, 2))
+    return scores
+
+
+@tf.function
+def segmentation_operator(model: Callable,
+                          inputs: tf.Tensor,
+                          targets: tf.Tensor) -> tf.Tensor:
+    """
+    Compute the segmentation score for a batch of samples.
+
+    Parameters
+    ----------
+    model
+        Model used for computing predictions.
+    inputs
+        Input samples to be explained.
+    targets
+        One-hot encoded labels or regression target (e.g {+1, -1}), one for each sample.
+
+    Returns
+    -------
+    scores
+        Segmentation scores computed.
+    """
+    scores = tf.reduce_sum(model(inputs) * targets, axis=(1, 2, 3))
+    return scores
+
+
+def get_gradient_of_operator(operator):
+    """
+    Get the gradient of an operator.
+
+    Parameters
+    ----------
+    operator
+        Operator to compute the gradient of.
+
+    Returns
+    -------
+    gradient
+        Gradient of the operator.
+    """
+    @tf.function
+    def gradient(model, inputs, targets):
+        with tf.GradientTape() as tape:
+            tape.watch(inputs)
+            scores = operator(model, inputs, targets)
+
+        return tape.gradient(scores, inputs)
+
+    return gradient
+
+
+def operator_batching(operator: Callable) -> tf.Tensor:
+    """
+    Take care of batching an operator: (model, inputs, labels).
+
+    Parameters
+    ----------
+    operator
+        Any callable that take model, inputs and labels as parameters.
+
+    Returns
+    -------
+    batched_operator
+        Function that apply operator by batch.
+    """
+
+    def batched_operator(model, inputs, targets, batch_size=None):
```

Comment:
I think we could do something more efficient with `map` or `map_fn` rather than using a for loop

Reply:
I agree that `map_fn` and `vectorized_map` would help for parallelization, but why do you want that for inference (just increase batch_size)? ^^

Still, I think that could be an interesting tool if we had preprocessing / data augmentation!